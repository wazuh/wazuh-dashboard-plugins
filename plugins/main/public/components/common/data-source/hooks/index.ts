export * from './use-data-source';