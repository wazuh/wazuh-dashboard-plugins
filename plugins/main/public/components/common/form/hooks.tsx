import { useState, useRef } from 'react';
import { isEqual, get, set, cloneDeep } from 'lodash';
import { EpluginSettingType } from '../../../../common/constants';
import {
  CustomSettingType,
  FormConfiguration,
  SettingTypes,
  UseFormReturn,
} from './types';

interface IgetValueFromEventType {
  [key: string]: (event: any) => any;
}

/**
 * Returns the value of the event according to the type of field
 * When the type is not found, it returns the value defined in the default key
 *
 * @param event
 * @param type
 * @returns event value
 */
function getValueFromEvent(
  event: any,
  type: SettingTypes | CustomSettingType,
): any {
  return (getValueFromEventType[type] || getValueFromEventType.default)(event);
}

const getValueFromEventType: IgetValueFromEventType = {
  [EpluginSettingType.switch]: (event: any) => event.target.checked,
  [EpluginSettingType.editor]: (value: any) => value,
  [EpluginSettingType.filepicker]: (value: any) => value,
  [EpluginSettingType.select]: (event: any) => event.target.value,
  [EpluginSettingType.text]: (event: any) => event.target.value,
  [EpluginSettingType.textarea]: (event: any) => event.target.value,
  [EpluginSettingType.number]: (event: any) => event.target.value,
  [EpluginSettingType.password]: (event: any) => event.target.value,
  custom: (event: any) => event.target.value,
  default: (event: any) => event.target.value,
};

export function getFormFields(fields) {
  return Object.entries(fields).reduce(
    (accum, [fieldKey, fieldConfiguration]) => {
      return {
        ...accum,
        [fieldKey]: {
          ...(fieldConfiguration.type === 'arrayOf'
            ? {
                fields: fieldConfiguration.initialValue.map(item =>
                  getFormFields(
                    Object.entries(fieldConfiguration.fields).reduce(
                      (accum, [key]) => ({
                        ...accum,
                        [key]: {
                          initialValue: item[key],
                          currentValue: item[key],
                          defaultValue: fieldConfiguration?.defaultValue,
                        },
                      }),
                      {},
                    ),
                  ),
                ),
              }
            : {
                currentValue: fieldConfiguration.initialValue,
                initialValue: fieldConfiguration.initialValue,
                defaultValue: fieldConfiguration?.defaultValue,
              }),
        },
      };
    },
    {},
  );
}

export function enhanceFormFields(
  formFields,
  {
    fields,
    references,
    setState,
    pathFieldParent = [],
    pathFormFieldParent = [],
  },
) {
  return Object.entries(formFields).reduce(
    (accum, [fieldKey, { currentValue: value, ...restFieldState }]) => {
      // Define the path to fields object
      const pathField = [...pathFieldParent, fieldKey];
      // Define the path to the form fields object
      const pathFormField = [...pathFormFieldParent, fieldKey];
      // Get the field form the fields
      const field = get(fields, pathField);

      return {
        ...accum,
        [fieldKey]: {
          ...(field.type === 'arrayOf'
            ? {
                type: field.type,
                fields: (() => {
                  return restFieldState.fields.map((fieldState, index) =>
                    enhanceFormFields(fieldState, {
                      fields,
                      references,
                      setState,
                      pathFieldParent: [...pathField, 'fields'],
                      pathFormFieldParent: [...pathFormField, 'fields', index],
                    }),
                  );
                })(),
                addNewItem: () => {
                  setState(state => {
                    const _state = get(state, [...pathField, 'fields']);
                    const newstate = set(
                      state,
                      [...pathField, 'fields', _state.length],
                      Object.entries(field.fields).reduce(
                        (accum, [key, { defaultValue }]) => ({
                          ...accum,
                          [key]: {
                            currentValue: cloneDeep(defaultValue),
                            initialValue: cloneDeep(defaultValue),
                            defaultValue: cloneDeep(defaultValue),
                          },
                        }),
                        {},
                      ),
                    );
                    return cloneDeep(newstate);
                  });
                },
              }
            : {
                ...field,
                ...restFieldState,
                type: field.type,
                value,
                changed: !isEqual(restFieldState.initialValue, value),
                error: field?.validate?.(value),
                setInputRef: (reference: any) => {
                  set(references, pathFormField, reference);
                },
                inputRef: get(references, pathFormField),
                onChange: (event: any) => {
                  const inputValue = getValueFromEvent(event, field.type);
                  const currentValue =
                    field?.transformChangedInputValue?.(inputValue) ??
                    inputValue;
                  setState(state => {
                    const newState = set(
                      cloneDeep(state),
                      [...pathFormField, 'currentValue'],
                      currentValue,
                    );
                    return newState;
                  });
                },
              }),
        },
      };
    },
    {},
  );
}

export function mapFormFields(
  {
    formDefinition,
    formState,
    pathFieldFormDefinition = [],
    pathFormState = [],
  },
  callbackFormField,
) {
  return Object.entries(formState).reduce((accum, [key, value]) => {
    const pathField = [...pathFieldFormDefinition, key];
    const fieldDefinition = get(formDefinition, pathField);
    return {
      ...accum,
      [key]:
        fieldDefinition.type === 'arrayOf'
          ? {
              fields: value.fields.map((valueField, index) =>
                mapFormFields(
                  {
                    formDefinition,
                    formState: valueField,
                    pathFieldFormDefinition: [...pathField, 'fields'],
                    pathFormState: [
                      ...[...pathFormState, key],
                      'fields',
                      index,
                    ],
                  },
                  callbackFormField,
                ),
              ),
            }
          : callbackFormField?.(value, key, {
              formDefinition,
              formState,
              pathFieldFormDefinition,
              pathFormState: [...pathFormState, key],
              fieldDefinition,
            }),
    };
  }, {});
}

export const useForm = (fields: FormConfiguration): UseFormReturn => {
  const [formFields, setFormFields] = useState<{
    [key: string]: { currentValue: any; initialValue: any };
  }>(getFormFields(fields));

  const fieldRefs = useRef<{ [key: string]: any }>({});

  const enhanceFields = enhanceFormFields(formFields, {
    fields,
    references: fieldRefs.current,
    setState: setFormFields,
    pathFieldParent: [],
    pathFormFieldParent: [],
  });

  const { changed, errors } = (() => {
    const result = {
      changed: {},
      errors: {},
    };
    mapFormFields(
      {
        formDefinition: fields,
        formState: enhanceFields,
        pathFieldFormDefinition: [],
        pathFormState: [],
      },
      ({ changed, error, value }, _, { pathFormState, fieldDefinition }) => {
        changed &&
          (result.changed[pathFormState] =
            fieldDefinition?.transformChangedOutputValue?.(value) ?? value);
        error && (result.errors[pathFormState] = error);
      },
    );
    return result;
  })();

  function undoChanges() {
    setFormFields(state =>
      mapFormFields(
        {
          formDefinition: fields,
          formState: state,
          pathFieldFormDefinition: [],
          pathFormState: [],
        },
        state => ({ ...state, currentValue: state.initialValue }),
      ),
    );
  }

  function doChanges() {
    setFormFields(state =>
      mapFormFields(
        {
          formDefinition: fields,
          formState: state,
          pathFieldFormDefinition: [],
          pathFormState: [],
        },
        state => ({ ...state, initialValue: state.currentValue }),
      ),
    );
  }

  function forEach(callback) {
    return mapFormFields(
      {
        formDefinition: fields,
        formState: enhanceFields,
        pathFieldFormDefinition: [],
        pathFormState: [],
      },
      callback,
    );
  }

  return {
    fields: enhanceFields,
    changed,
    errors,
    undoChanges,
    doChanges,
    forEach,
  };
};
