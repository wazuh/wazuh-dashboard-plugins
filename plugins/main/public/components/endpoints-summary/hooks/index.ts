export { useGetTotalAgents } from './agents';
export { useGetGroups } from './groups';
