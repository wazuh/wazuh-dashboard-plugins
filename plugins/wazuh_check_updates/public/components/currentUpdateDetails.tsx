import React from 'react';
import { EuiCallOut, EuiHeaderLink } from '@elastic/eui';
import { Update } from '../../common/types';
import { FormattedMessage } from '@osd/i18n/react';

export interface CurrentUpdateDetailsProps {
  currentUpdate?: Update;
}

export const CurrentUpdateDetails = ({ currentUpdate }: CurrentUpdateDetailsProps) => {
  if (!currentUpdate) return null;

  const currentRelease = `${currentUpdate?.semver.mayor}.${currentUpdate?.semver.minor}`;
  const releaseNotesUrl = `https://documentation.wazuh.com/${currentRelease}/release-notes/release-${currentUpdate?.semver.mayor}-${currentUpdate?.semver.minor}-${currentUpdate?.semver.patch}.html`;
  const upgradeGuideUrl = `https://documentation.wazuh.com/${currentRelease}/upgrade-guide/index.html`;

  return (
    <EuiCallOut title="Proceed with caution!" color="warning" iconType="help">
      <p>{currentUpdate.title}</p>
      <EuiHeaderLink href={releaseNotesUrl} isActive target="_blank">
        {
          <FormattedMessage
            id="wazuhCheckUpdates.currentUpdateDetails.releaseNotesLink"
            defaultMessage="Release notes"
          />
        }
      </EuiHeaderLink>
      <EuiHeaderLink href={upgradeGuideUrl} isActive target="_blank">
        {
          <FormattedMessage
            id="wazuhCheckUpdates.currentUpdateDetails.upgradeGuideLink"
            defaultMessage="Upgrade guide"
          />
        }
      </EuiHeaderLink>
    </EuiCallOut>
  );
};
