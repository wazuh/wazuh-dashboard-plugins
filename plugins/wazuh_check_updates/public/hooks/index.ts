export { useAvailableUpdates } from './useAvailableUpdates';
export { useUserPreferences } from './useUserPreferences';
