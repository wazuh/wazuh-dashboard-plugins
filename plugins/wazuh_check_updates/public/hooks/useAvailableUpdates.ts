import { useState, useEffect } from 'react';
import { AvailableUpdates } from '../../common/types';
import { routes } from '../../common/index';
import { getCore } from '../plugin-services';

export const useAvailableUpdates = () => {
  const [availableUpdates, setAvailableUpdates] = useState<AvailableUpdates>({
    mayor: [],
    minor: [],
    patch: [],
  });

  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const refreshAvailableUpdates = (forceUpdate = false) => {
    (async () => {
      try {
        setIsLoading(true);
        const response = await getCore().http.get(`${routes.checkUpdates}`, {
          query: {
            checkAvailableUpdates: forceUpdate,
          },
        });
        setAvailableUpdates(response);
        setError(undefined);
      } catch (error: any) {
        setError(error);
      } finally {
        setIsLoading(false);
      }
    })();
  };

  useEffect(() => {
    refreshAvailableUpdates();
  }, []);

  return { isLoading, availableUpdates, refreshAvailableUpdates, error };
};
