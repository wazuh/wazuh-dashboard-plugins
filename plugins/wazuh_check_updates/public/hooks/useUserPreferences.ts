import { useState, useEffect } from 'react';
import { UserPreferences } from '../../common/types';
import { getCore } from '../plugin-services';
import { routes } from '../../common/index';

export const useUserPreferences = () => {
  const [userPreferences, setUserPreferences] = useState<UserPreferences>({});
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState();

  const refreshUserPreferences = () => {
    (async () => {
      try {
        setIsLoading(true);
        const response = await getCore().http.get(routes.userPreferences);
        setUserPreferences(response);
        setError(undefined);
      } catch (error: any) {
        setError(error);
      } finally {
        setIsLoading(false);
      }
    })();
  };

  useEffect(() => {
    refreshUserPreferences();
  }, []);

  const updateUserPreferences = async (userPreferences: UserPreferences) => {
    try {
      setIsLoading(true);
      await getCore().http.patch(routes.userPreferences, {
        body: JSON.stringify(userPreferences),
      });
      setUserPreferences(userPreferences);
      setIsLoading(false);
    } catch (error: any) {
      setError(error);
      setIsLoading(false);
    }
  };

  return {
    isLoading,
    userPreferences,
    updateUserPreferences,
    error,
  };
};
