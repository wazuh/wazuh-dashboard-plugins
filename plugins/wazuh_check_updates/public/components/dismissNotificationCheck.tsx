import { EuiCheckbox } from '@elastic/eui';
import React, { useState } from 'react';
import { FormattedMessage } from '@osd/i18n/react';
import { useUserPreferences } from '../hooks';

export interface DismissNotificationCheckProps {
  userId: string;
}

export const DismissNotificationCheck = ({ userId }: DismissNotificationCheckProps) => {
  const [dismissFutureUpdates, setDismissFutureUpdates] = useState(false);

  const { userPreferences, error, isLoading, updateUserPreferences } = useUserPreferences(userId);

  if (error) {
    console.log(error);
    return null;
  }

  if (isLoading) return null;

  const handleOnChange = (checked: boolean) => {
    updateUserPreferences({ hide_update_notifications: checked });
    setDismissFutureUpdates(true);
  };

  return (
    <EuiCheckbox
      id="check-dismiss"
      label={
        <FormattedMessage
          id="wazuhCheckUpdates.updatesNotification.dismissCheckText"
          defaultMessage="I don't want to know about future releases"
        />
      }
      defaultChecked={userPreferences.hide_update_notifications}
      checked={dismissFutureUpdates}
      onChange={(e) => handleOnChange(e.target.checked)}
    />
  );
};
