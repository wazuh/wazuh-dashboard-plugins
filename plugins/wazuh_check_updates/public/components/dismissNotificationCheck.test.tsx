import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { CurrentUpdateDetails } from './currentUpdateDetails';
import { TestProviders } from '../test/test-utils';
import { DismissNotificationCheck } from './dismissNotificationCheck';
import { useUserPreferences } from '../hooks';

const mockedUseUserPreferences = useUserPreferences as jest.Mock;
jest.mock('../hooks/useUserPreferences');

describe('DismissNotificationCheck component', () => {
  test('should render the check', () => {
    mockedUseUserPreferences.mockImplementation(() => ({
      isLoading: false,
      userPreferences: { hide_update_notifications: false, last_dismissed_update: 'v4.2.1' },
    }));

    const { container, getByText } = render(
      <TestProviders>
        <DismissNotificationCheck />
      </TestProviders>
    );

    expect(container).toMatchSnapshot();

    const elementWithTag = getByText("I don't want to know about future releases");
    expect(elementWithTag).toBeInTheDocument();
  });

  test('should return null when there is an error', () => {
    mockedUseUserPreferences.mockImplementation(() => ({
      error: 'Error',
    }));

    const { container } = render(<CurrentUpdateDetails />);

    expect(container).toMatchSnapshot();

    const firstChild = container.firstChild;
    expect(firstChild).toBeNull();
  });
});
