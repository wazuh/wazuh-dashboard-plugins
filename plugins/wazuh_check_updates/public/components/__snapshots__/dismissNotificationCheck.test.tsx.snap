// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DismissNotificationCheck component should render the check 1`] = `
<div>
  <div
    class="euiCheckbox"
  >
    <input
      class="euiCheckbox__input"
      id="check-dismiss"
      type="checkbox"
    />
    <div
      class="euiCheckbox__square"
    />
    <label
      class="euiCheckbox__label"
      for="check-dismiss"
    >
      <span>
        I don't want to know about future releases
      </span>
    </label>
  </div>
</div>
`;

exports[`DismissNotificationCheck component should return null when is loading 1`] = `<div />`;

exports[`DismissNotificationCheck component should return null when there is an error 1`] = `<div />`;
